src/main/java/org/graylog/integrations/inputs/paloalto/PaloAltoTypeParser.java

```diff
@@ -41,68 +45,62 @@ public PaloAltoTypeParser(PaloAltoMessageTemplate messageTemplate, PaloAltoMessa
-            } catch (IndexOutOfBoundsException e) {
-                // Skip fields at indexes that do not exist.
-                LOG.trace("A [{}] field does not exist at index [{}]", messageType.toString(), field.position());
-                continue;
-            }
+        int fieldIndex = 0;
+        int templateIndex = 0;
 
-            Object value = null;
+        while (fieldIndex < fields.size() && templateIndex < templateFields.size()) {
+            String rawValue = fields.get(fieldIndex);
+            PaloAltoFieldTemplate template = templateFields.get(templateIndex);
 
-            switch (field.fieldType()) {
-                case STRING:
-                    // Handle quoted values.
+            if (fieldIndex < template.position()) {
+                fieldIndex++;
+                continue;
+            } else if (fieldIndex == template.position()) {
+                Object value = rawValue;
+
+                if (template.fieldType() == STRING) {
                     if (rawValue.startsWith("\"") && rawValue.endsWith("\"")) {
-                        rawValue = rawValue.substring(1, rawValue.length() - 1);
+                        value = rawValue.substring(1, rawValue.length() - 1);
                     }
-
-                    value = rawValue;
-                    break;
-                case LONG:
-                    if (!Strings.isNullOrEmpty(rawValue)) {
-                        try {
-                            value = Long.valueOf(rawValue);
-                        } catch (NumberFormatException e) {
-                            LOG.error("[{}] is an invalid LONG value for the [{}] [{}] field", rawValue, messageType, field.field() );
-                            continue;
-                        }
-                    } else {
-                        value = 0L;
+                } else if (template.fieldType() == LONG) {
+                    try {
+                        value = Strings.isNullOrEmpty(rawValue) ? 0L : Long.valueOf(rawValue);
+                    } catch (NumberFormatException ex) {
+                        LOG.error("Error parsing field {}, {} is not a valid numeric value", template.field(), rawValue);
+                        value = null;
                     }
-                    break;
-                case BOOLEAN:
-                    if (!Strings.isNullOrEmpty(rawValue)) {
-                        value = Boolean.valueOf(rawValue);
+                } else if (template.fieldType() == BOOLEAN) {
+                    value = Boolean.valueOf(rawValue);
+                } else {
+                    LOG.warn("Unrecognized data type [{}] for field [{}], handling as STRING",
+                            template.fieldType(), template.field());
+                }
+
+                // Handling of duplicate keys
+                if (fieldMap.containsKey(template.field())) {
+                    if (Strings.isNullOrEmpty(rawValue.trim()) || value.equals(fieldMap.get(template.field()))) {
```

Comment:
The IDE warns about a possible NPE with `value.equals(...)`. Not sure if that can happen in practice.

Reply:
`value` could be null if `rawValue` was null, in which case the `Strings.isNullOrEmpty(..)` check would catch it.  It could also be null if we caught a NumberFormatException trying to parse a Long.  I'll add a null check just to be safe.